File: leetcode/141-linked-list-cycle.js
// https://leetcode.com/problems/linked-list-cycle/ (top-interview-questions)


// Definition for singly-linked list.
function ListNode(val) {
  this.val = val;
  this.next = null;
}

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
  if (head === null || head.next === null) return false;
  
  let tortoise = head;
  let hare = head.next.next;

  while (hare) {
    if (hare === tortoise) return true;
    tortoise = tortoise.next;
    hare = hare.next?.next;
  }

  return false;
};

// bad solution with O(n) space complexity
var hasCycle = function(head) {
  if (head === null || head.next === null) return false;
  
  const nodes = new Set();
  
  let node = head;

  while (node) {
    nodes.add(node);
    if (nodes.has(node.next)) return true;
    node = node.next;
  }

  return false;
}

File: leetcode/141-linked-list-cycle.py
# https://leetcode.com/problems/linked-list-cycle/ (top-interview-questions)

from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        tortoise, hare = head, head.next.next if (head and head.next) else None
        while hare is not None:
            if tortoise == hare: return True
            tortoise = tortoise.next
            hare = hare.next.next if hare.next else None
        return False

# floyd's tortoise and hare algorithm (cycle detection)